### PR TITLE
Fix locale issues with Mantid Tables

### PR DESCRIPTION
**Pull request by samueljackson92**
PR #20401 Fixed truncation issues with table workspaces but reintroduced issues with locales into Mantid. This PR should hopefully fix both.

**To test:**

Here's script to produce a table workspace. Make sure the "hello world" does not get truncated at the first space or converted to a 0.

Change your locale to something that isn't english and check the data is displayed correctly.

```python
ws = CreateEmptyTableWorkspace()
ws.addColumn("str", "value")
ws.addColumn("double", "v")
ws.addColumn("int", "v2")
ws.addColumn("V3D", "v3")
ws.addColumn("double", "v4")
ws.addRow(["hello world", 1000000.0, 123, V3D(0,1,0), 3.14159])
```

_There is no issue associated with this PR_

**Release Notes** 
*Does not need to be in the release notes.*

---

#### Reviewer ####

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):

##### Code Review #####

- [x] Is the code of an acceptable quality?
- [x] Does the code conform to the [coding standards](http://www.mantidproject.org/Coding_Standards)?
- [ ] Are the unit tests small and test the class in isolation?
- [ ] If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- [x] Do changes function as described? Add comments below that describe the tests performed?
- [x] Do the changes handle unexpected situations, e.g. bad input?
- [ ] Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: MantidPlot/src/Mantid/MantidTable.cpp [hunk 1]
```diff
@@ -294,14 +297,15 @@ void MantidTable::cellEdited(int row, int col) {
   // Have the column convert the text to a value internally
   int index = row;
   std::istringstream textStream(text);
-  const std::locale systemLocale("");
-  textStream.imbue(systemLocale);
-  c->read(index, textStream.str());
+  const auto localName = locale().name().toStdString();
+  std::locale applicationLocale(localName.c_str());
```
[the remark this answers is not in the document]
Comment (samueljackson92) [reply MantidPlot/src/Mantid/MantidTable.cpp]: Urgh how disgusting. Thanks for showing me how to fix it though @soininen.  I'll play around with it and see if there's a nicer way.

[the remark this answers is not in the document]
Comment (VickieLynch) [reply MantidPlot/src/Mantid/MantidTable.cpp]: Just tested this and it fixes ShowData errors for PeaksWorkspaces.  SNS users were asking about this error yesterday, so hope it is in nightly soon.

**Comment by samueljackson92** [comment]
@soininen As far as I can see there isn't a way to get `QLocale` and `std::locale` to work nicely together. This fix isn't great, but it does allow the locale support to work in the table view and keeps the format read by the `TableColumn` in the standard locale.

**Comment by samueljackson92** [comment]
@soininen I tested with the updated script in the description and changing my settings in `View -> Preferences -> Numeric Format`. I see to get what you expect now and the data also changes if I edit a cell's contents.

**Review by VickieLynch — APPROVED** [review]
bug with Show Data for PeaksWorkspaces fixed

**Comment by samueljackson92** [comment]
@VickieLynch did you check the changes with a locale other than `en_US` or `en_UK`?

**Comment by VickieLynch** [comment]
@samueljackson92 No

**Review by soininen — APPROVED** [review]
Now strings don't get truncated and numeric cells work with the locale settings at ILL. The conversion from Qt to C locale is a bit hacky, but I guess there is no easy way around it. It is better solution than the original which was happily mixing the Qt and C locales for reading and writing table data.

**Comment by samueljackson92** [comment]
@soininen It also avoids using an exception for [flow control](http://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Re-errors).

**Pull request merged**